File: app/lib/dream/deckContentCanonical.py
# [Input] Fixed JSON-only aggregate facts or memory JSON, never SQL/paths/credentials.
# [Output] Exact existing Python deck-content/v1 canonical bytes/hash/diff, with safe failure.
# [Pos] Fixed Admin stdlib codec for domain facts; no SQL, network, service or Runtime execution.
# [Sync] 2026-09-15: preserve canonical bytes, claim equality, memory dictionaries and three capability-set decoders.
import hashlib
import json
import sys


def canonical(value, *, allow_nan=True):
    return json.dumps(value, ensure_ascii=False, allow_nan=allow_nan, separators=(",", ":"), sort_keys=True)


def json_value(raw):
    if raw is None:
        return None
    try:
        return json.loads(raw)
    except (ValueError, TypeError):
        return raw


def change(scope, label, change_type="modified", fields=None):
    return {"scope": scope, "label": label, "change_type": change_type, "fields": fields or []}


def diff(base, current):
    if base is None:
        changes = [change("deck", "Deck 基础信息", "added"), change("agent_type", "Agent 类型", "added")]
        for key, label in (("agents", "Agents"), ("claude_plugins", "Claude 插件"), ("runtime_binding", "运行绑定")):
            if current.get(key):
                changes.append(change(key, label, "added"))
        return changes
    changes = []
    old = base.get("deck") if isinstance(base.get("deck"), dict) else {}
    new = current.get("deck") if isinstance(current.get("deck"), dict) else {}
    fields = sorted(key for key in set(old) | set(new) if old.get(key) != new.get(key))
    if fields:
        changes.append(change("deck", "Deck 基础信息", fields=fields))
    if base.get("agent_type") != current.get("agent_type"):
        changes.append(change("agent_type", "Agent 类型", fields=["agent_type"]))
    for key, label in (("agents", "Agents"), ("claude_plugins", "Claude 插件"), ("runtime_binding", "运行绑定")):
        if base.get(key) != current.get(key):
            changes.append(change(key, label))
    return changes



def confirmation_envelope(raw_parts, actor):
    try:
        parts = json.loads(raw_parts)
        if not isinstance(parts, list) or len(parts) != 1 or not isinstance(parts[0], dict):
            return {"status": "invalid"}
        if parts[0].get("type") != "text" or not isinstance(parts[0].get("text"), str):
            return {"status": "invalid"}
        envelope = json.loads(parts[0]["text"])
        if not isinstance(envelope, dict) or envelope.get("kind") != "story-workspace-dream-confirmation":
            return {"status": "invalid"}
        command = envelope.get("command")
        if not isinstance(command, dict) or not isinstance(actor, str) or not actor:
            return {"status": "invalid"}
        run, thread, key = (command.get(name) for name in ("storyWorkspaceRunId", "threadId", "idempotencyKey"))
        if not all(isinstance(value, str) and bool(value) for value in (run, thread, key)):
            return {"status": "invalid"}
        command_json = canonical(command, allow_nan=False)
        fingerprint_input = canonical({"actor": actor, "command": command}, allow_nan=False)
        message_input = canonical({"actor": actor, "storyWorkspaceRunId": run, "idempotencyKey": key}, allow_nan=False)
        return {"status": "valid", "story_workspace_run_id": run, "thread_id": thread, "idempotency_key": key,
                "command_json": command_json, "command_fingerprint": "sha256:" + hashlib.sha256(fingerprint_input.encode()).hexdigest(),
                "message_id": "dream_confirm_" + hashlib.sha256(message_input.encode()).hexdigest(),
                "parts_canonical_json": canonical(parts, allow_nan=False)}
    except (ValueError, TypeError):
        return {"status": "invalid"}


def confirmation_claims(left_raw, right_raw):
    try:
        left, right = json.loads(left_raw), json.loads(right_raw)
        if not isinstance(left, dict) or not isinstance(right, dict):
            return {"equal": False}
        left.pop("dispatch_claim_lease_until", None)
        right.pop("dispatch_claim_lease_until", None)
        return {"equal": left == right}
    except (ValueError, TypeError):
        return {"equal": False}


def execute(request):
    if request.get("action") == "plugin-capability-sets":
        try:
            parsed = json.loads(request["json_text"] or "[]")
        except (ValueError, TypeError):
            parsed = None
        context = {item.strip() for item in parsed if isinstance(item, str) and item.strip()} if isinstance(parsed, list) else set()
        approved = set(parsed) if isinstance(parsed, list) and all(isinstance(item, str) for item in parsed) else set()
        try:
            preflight = {item for item in set(parsed) if isinstance(item, str)}
        except TypeError:
            preflight = set()
        return {"context_names": sorted(context), "approved_names": sorted(approved), "preflight_names": sorted(preflight)}
    if request.get("action") == "memory-inspect":
        return {"is_object": isinstance(json_value(request["json_text"]), dict)}
    if request.get("action") == "confirmation-envelope":
        return confirmation_envelope(request["raw_parts_json"], request["actor_id"])
    if request.get("action") == "confirmation-claims":
        return confirmation_claims(request["stored_metadata_json"], request["incoming_metadata_json"])
    if request.get("action") == "canonical":
        value = json.loads(request["json_text"])
        text = canonical(value, allow_nan=False)
        return {"canonical_json": text, "content_hash": "sha256:" + hashlib.sha256(text.encode()).hexdigest()}
    if request.get("action") == "memory":
        left = json_value(request["left"])
        right = json_value(request["right"])
        return {"equal": canonical(left) == canonical(right), "canonical_json": canonical(right)}
    if request.get("action") == "snapshot":
        current = request["facts"]
        for agent in current["agents"]:
            agent["memory_workspace_config"] = json_value(agent.pop("memory_workspace_config_json"))
        base = json_value(request["base_json"])
        if base is not None and not isinstance(base, dict):
            raise ValueError("Invalid snapshot")
        text = canonical(current)
        digest = lambda value: "sha256:" + hashlib.sha256(canonical(value).encode()).hexdigest()
        changes = diff(base, current)
        return {"canonical_json": text, "content_hash": digest(current), "changes": changes, "no_changes": not changes or (base is not None and digest(base) == digest(current))}
    raise ValueError("Unknown action")


if __name__ == "__main__":
    try:
        json.dump(execute(json.load(sys.stdin)), sys.stdout, ensure_ascii=True, separators=(",", ":"))
    except Exception:
        sys.stderr.write("DECK_CANONICAL_FAILED\n")
        sys.exit(1)

File: app/lib/dream/deckContentCanonical.ts
// [Input] Admin-owned aggregate facts with raw memory/base JSON and a configured canonical-operation deadline.
// [Output] Python-compatible v1 hash/diff; never parses legacy numeric payload through JS Number.
// [Pos] Fixed local pure helper, no shell, DB, network, user executable or Runtime configuration.
// [Sync] 2026-09-15: preserve raw provenance/memory/capability decoding with bounded safe transport.
import { spawn } from "node:child_process";
import { resolve } from "node:path";
import { z } from "zod";
import { AuthBoundaryError, requiredAuthValue } from "../auth/config";
import { deckVersionChangeDto } from "./deckVoiceDto";
const snapshotResult = z.strictObject({ canonical_json: z.string(), content_hash: z.string().regex(/^sha256:[0-9a-f]{64}$/), changes: z.array(deckVersionChangeDto), no_changes: z.boolean() });
const memoryResult = z.strictObject({ canonical_json: z.string(), equal: z.boolean() });
async function invoke(input: unknown) {
  const timeout = Number(requiredAuthValue("DREAM_DOMAIN_CANONICAL_TIMEOUT_MS"));
  if (!Number.isSafeInteger(timeout) || timeout <= 0) throw new AuthBoundaryError("DECK_CANONICAL_NOT_CONFIGURED");
  return new Promise<unknown>((done, reject) => {
    const child = spawn("python3", ["-I", "-S", resolve(process.cwd(), "app/lib/dream/deckContentCanonical.py")], { shell: false, env: { PATH: process.env.PATH } as unknown as NodeJS.ProcessEnv, stdio: "pipe" });
    const chunks: Buffer[] = []; let failed = false;
    const failure = () => { if (!failed) { failed = true; clearTimeout(timer); reject(new AuthBoundaryError("DECK_CANONICAL_UNAVAILABLE")); } };
    const timer = setTimeout(() => { child.kill(); failure(); }, timeout);
    child.stdout.on("data", chunk => chunks.push(chunk)); child.stderr.resume();
    child.on("error", failure); child.stdin.on("error", failure);
    child.on("close", code => { clearTimeout(timer); if (failed) return; if (code !== 0) return failure(); try { done(JSON.parse(Buffer.concat(chunks).toString("utf8"))); } catch { failure(); } });
    child.stdin.end(JSON.stringify(input));
  });
}
export async function analyzeDeckSnapshot(facts: unknown, baseJson: string | null) { return snapshotResult.parse(await invoke({ action: "snapshot", facts, base_json: baseJson })); }
export async function canonicalMemory(left: string | null, right: string | null) { return memoryResult.parse(await invoke({ action: "memory", left, right })); }

// Internal business provenance codec; no HTTP operation or SQL selector is exposed.
export async function canonicalBusinessJson(rawBusinessJson: string) { return z.strictObject({canonical_json:z.string(),content_hash:z.string().regex(/^sha256:[0-9a-f]{64}$/)}).parse(await invoke({action:"canonical",json_text:rawBusinessJson})); }

const confirmationEnvelopeResult = z.discriminatedUnion("status", [
 z.strictObject({status:z.literal("invalid")}),
 z.strictObject({status:z.literal("valid"),story_workspace_run_id:z.string().min(1),thread_id:z.string().min(1),idempotency_key:z.string().min(1),command_json:z.string(),command_fingerprint:z.string().regex(/^sha256:[a-f0-9]{64}$/),message_id:z.string().regex(/^dream_confirm_[a-f0-9]{64}$/),parts_canonical_json:z.string()}),
]);
export async function analyzeConfirmationEnvelope(rawPartsJson:string,actorId:string){return confirmationEnvelopeResult.parse(await invoke({action:"confirmation-envelope",raw_parts_json:rawPartsJson,actor_id:actorId}));}
export async function compareConfirmationClaims(storedMetadataJson:string,incomingMetadataJson:string){return z.strictObject({equal:z.boolean()}).parse(await invoke({action:"confirmation-claims",stored_metadata_json:storedMetadataJson,incoming_metadata_json:incomingMetadataJson}));}

// Preserve original Python dictionary detection without decoding stored numeric JSON through JS Number.
export async function inspectMemoryConfig(rawMemoryJson:string|null){return z.strictObject({is_object:z.boolean()}).parse(await invoke({action:"memory-inspect",json_text:rawMemoryJson}));}

// Three original callers intentionally decode approved JSON differently; use
// Python for legacy nonfinite values and hashable set/string/dictionary rules.
export async function inspectPluginCapabilitySets(rawJson: string | null) {
  return z.strictObject({ context_names: z.array(z.string()), approved_names: z.array(z.string()), preflight_names: z.array(z.string()) })
    .parse(await invoke({ action: "plugin-capability-sets", json_text: rawJson }));
}
